src/Microsoft.AspNet.Mvc.Razor.Host/Directives/CodeTreeCache.cs

```diff
@@ -0,0 +1,66 @@
+﻿// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Concurrent;
+using Microsoft.AspNet.FileProviders;
+using Microsoft.AspNet.Razor.Generator.Compiler;
+using Microsoft.Framework.Expiration.Interfaces;
+
+namespace Microsoft.AspNet.Mvc.Razor.Directives
+{
+    /// <summary>
+    /// Default implementation of <see cref="ICodeTreeCache"/>.
+    /// </summary>
+    public class CodeTreeCache : ICodeTreeCache
+    {
+        private readonly IFileProvider _fileProvider;
+        private readonly ConcurrentDictionary<string, CodeTreeCacheEntry> _codeTreeCache;
+
+        /// <summary>
+        /// Initializes a new instance of <see cref="CodeTreeCache"/>.
+        /// </summary>
+        /// <param name="fileProvider">The application's <see cref="IFileProvider"/>.</param>
+        public CodeTreeCache(IFileProvider fileProvider)
+        {
+            _fileProvider = fileProvider;
+            _codeTreeCache = new ConcurrentDictionary<string, CodeTreeCacheEntry>(StringComparer.Ordinal);
+        }
+
+        /// <inheritdoc />
+        public CodeTree GetOrAdd([NotNull] string pagePath,
+                                 [NotNull] Func<IFileInfo, CodeTree> getCodeTree)
+        {
+            CodeTreeCacheEntry cacheEntry;
+            if (_codeTreeCache.TryGetValue(pagePath, out cacheEntry))
```

Comment:
Depends on how much stuff is being cached here and he eviction policies. If this stuff stays in memory long term because of that (even when file is not open anymore in the editor) OR long term in the runtime, we should reconsider this approach.